ts/deploy.ts

```diff
@@ -20,32 +17,37 @@ import {
     SpokeRegistryFactory,
     VaultFactory,
     WithdrawManagerFactory,
-    Create2TransferFactory
+    Create2TransferFactory,
+    DeployerFactory
 } from "../types/ethers-contracts";
 import { State } from "./state";
 import { merklise } from "./utils";
 import { BurnAuctionFactory } from "../types/ethers-contracts/BurnAuctionFactory";
 import { BurnAuction } from "../types/ethers-contracts/BurnAuction";
 import { ProofOfBurnFactory } from "../types/ethers-contracts/ProofOfBurnFactory";
 import { ProofOfBurn } from "../types/ethers-contracts/ProofOfBurn";
+import { getCreate2Address, id, keccak256 } from "ethers/lib/utils";
+import { Deployer } from "../types/ethers-contracts/Deployer";
 
-async function waitAndRegister(
-    contract: Contract,
-    name: string,
-    verbose: boolean,
-    nameRegistry?: NameRegistry,
-    registryKey?: string
-) {
-    await contract.deployed();
-    if (verbose) console.log("Deployed", name, "at", contract.address);
-    if (nameRegistry) {
-        if (!registryKey) throw Error(`Need registry key for ${name}`);
-        const tx = await nameRegistry.registerName(
-            registryKey,
-            contract.address
-        );
+class Create2Deployer {
+    public signer: Signer;
+    constructor(public deployerContract: Deployer) {
+        this.signer = this.deployerContract.signer;
+    }
+    public async deploy(
+        request: providers.TransactionRequest
+    ): Promise<string> {
+        const signerAddress = await this.signer.getAddress();
+        const salt = id(signerAddress);
+        const data = request.data as string;
```

Comment:
We are not able to resolve the circular dependency(two contracts' constructor depends on other's address) with CREATE2, because the CREATE2 address depends on the constructor args